Explore: Use SeriesData format for loki/logs

marefr opened the pull request:
**What this PR does / why we need it**:
This is the first step moving towards Explore supporting logs for additional datasources than Loki. In the first step we move all the log processing from Loki into Explore.

Summary of changes:
Make explore convert logs result returned from datasources to SeriesData, 
if needed, and for now convert SeriesData to LogsModel.
Loki datasource query now returns SeriesData and all
processing have been moved into explore instead.
Removed key from LogRowModel and use log row indexes as 
the unique key instead.
Removed id from LogsModel since it looks like it's not in use.

**Which issue(s) this PR fixes**:
Closes #16287

**Special notes for your reviewer**:
I removed `key` from `LogRowModel` and used index of each log row as key for rendering unique keys for `<LogRow />` component. To me this seems to work and makes the row model easier, but I may have missed some important aspect?

marefr committed:
explore: use SeriesData format for loki/logs

Make explore expect SeriesData to be returned as logs result
from datasources and for now convert SeriesData to LogsModel.
Loki datasource query now returns SeriesData and all
processing have been moved into explore instead.

marefr commented:
@ryantxu Please have a look at the changes if you're interested to see if this maps to your vision in referenced issue. This is the first step and there will probably be a lot of other changes needed, but now we make it easier for each datasource to return log data.

marefr commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -85,29 +77,24 @@ export class LokiDatasource {
+            seriesData.meta = {
+              search: queryTargets[i].regexp,
+              limit: this.maxLines,
+            };
```
For now added these properties into meta field.

torkelo replied:
think that should be a good use of the meta field

ryantxu replied:
+1 -- these also seem reasonable to put in the typescript defs rather than just `any`

marefr replied:
Fixed

marefr commented on public/app/plugins/datasource/loki/result_transformer.ts:
```diff
@@ -1,155 +1,16 @@
   return {
     labels,
-    fields: [{ name: 'ts', type: FieldType.time }, { name: 'line', type: FieldType.string }],
+    fields: [{ name: 'time', type: FieldType.time }, { name: 'message', type: FieldType.string }],
```
Decided to use time and message to make it more datasource agnostic. Also time is often used for metrics so thought that can be a good thing.

torkelo replied:
👍

ryantxu replied:
time is good -- but it should not matter what name you use since you the type is 'time' and on the other end you should use `getFirstTimeField`

marefr replied:
Have fixed this to be name-independent and use time and string field as requirement instead.

marefr commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+  const logSeries: SeriesData[] = [];
+
+  for (const series of seriesData) {
+    if (series.name !== 'logs') {
```
I guess we can skip this check since we're validating that a time and message column exists?

ryantxu replied:
no need for this -- it just makes it harder for other datasources to implement

marefr replied:
Removed

marefr reviewed (commented):
Some comments on some of the design decision I've made.

torkelo commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+    .reduce(
+      (acc: LogRowModel[], series: SeriesData, index: number) => [
+        ...acc,
+        ...series.rows.map(row => processLogSeriesRow(series, row, uniqueLabelsPerSeries[index])),
```
using map on rows is not ideal from a performance perspective, map is 100 times slower than a regular for lop, and there could be 100 to many thousands of rows.

torkelo replied:
For query result that returns many series this seems like it cause 2 array copy per series (as reduce returns a new array for each loop, copying results from the last). Looks a bit inefficient :)

marefr replied:
Have tried to fix this. Please review

torkelo commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+    .reverse()
+    .value();
+
+  const hasUniqueLabels = sortedRows && sortedRows.some(row => Object.keys(row.uniqueLabels).length > 0);
```
doing this in the main row loop would be preferable instead of looping in worst case a lot rows again (via callback fn). also why the sortedRows null check? how can it be null?

marefr replied:
Have tried to fix this. Please review

torkelo reviewed (commented):
Great work! So nice to see progress on this front!  Just some minor comments about the overuse of callback loops (reduce/maps etc) for arrays that can have thousands of entries.

ryantxu commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+
+    if (!hasFieldNamed(series, 'message')) {
+      continue;
+    }
```
Do we need to require that the name is 'message' can it just be the first/longest string?  or something like that?

marefr replied:
Fixed

ryantxu commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+  const timeEpochMs = time.valueOf();
+  const timeFromNow = time.fromNow();
+  const timeLocal = time.format('YYYY-MM-DD HH:mm:ss');
+  const logLevel = getLogLevel(message);
```
we don't need to solve this now, but it would be great to have the log level possibly come from the query:
`SELECT time,level,messsage FROM ...`

marefr replied:
Have added some logic for this, but probably needs some more love to support other field names than "loglevel".

ryantxu commented:
@marefr -- this looks really good.  thank you.

marefr committed:
Merge remote-tracking branch 'origin/master' into 16287_loki_series_data

marefr committed:
new SeriesFieldProcessor that encapsulates fields processing

marefr committed:
refactor logs processing

LogsSeriesFieldProcessor extends SeriesFieldProcessor and can validate
that series are valid for being visualized as logs. Valid log series
is now defined as having a time and a string field. Also, there now
support for checking if a series has a log level field.
The processing and sorting of logs have been refactored to hopefully
be more efficient than before.

marefr committed:
loki: remove usage of specific series and field names

marefr committed:
add optional fields to QueryResultMeta

marefr committed:
fix tests

marefr commented:
@ryantxu @torkelo thanks for the feedback :+1: Have pushed some changes now. I guess the biggest change is the `SeriesFieldsProcessor` which I did find handy when processing a lot of rows without needing to loop thru all the fields every time to find the correct field indices. 

Besides this have tried to make the log processing more efficient by using regular loops instead.

marefr commented on public/app/features/explore/LogsContainer.tsx:
```diff
@@ -78,7 +78,6 @@ export class LogsContainer extends PureComponent<LogsContainerProps> {
           data={logsResult}
           dedupedData={dedupedResult}
           exploreId={exploreId}
-          key={logsResult && logsResult.id}
```
Removed `id` from `LogsModel` since I don't think the key is needed here - to me should only be needed if you're rendering multiple `<Logs />` components.

marefr committed:
add tests and fix bugs

ryantxu commented on packages/grafana-ui/src/types/data.ts:
```diff
@@ -41,6 +47,10 @@ export interface Field {
+export interface IndexedField extends Field {
+  index: number;
+}
+
```
Can we move `IndexedField` to `processSeriesData.ts`?  I'm hoping `data.ts` just represents the core data exchange formats and the rest can be more efemeral/helpers

marefr replied:
Makes sense :+1: Will fix

ryantxu commented on packages/grafana-ui/src/utils/processSeriesData.ts:
```diff
@@ -212,3 +202,195 @@ export function sortSeriesData(data: SeriesData, sortIndex?: number, reverse = f
+  stringFieldIndices: number[];
+  numberFieldIndices: number[];
+  booleanFieldIndices: number[];
+  otherFieldIndices: number[];
```
rather than a new array for each type, what about:
```
fieldTypeIndices: { [type: FieldType]: number[] };
```

and than adding FieldType as a parameter to the getters below?  That feels less brittle as we add/refine the various field type options and approaches.

marefr replied:
Good point

ryantxu commented:
@marefr this is looking really good -- thanks.

The name on `SeriesFieldsProcessor` feels a bit off, and potentially in conflict with some series processing/transforming work I hope to standardize in #16756

As implemented it is a way to find field indexes based on name/type -- maybe `SeriesFieldLocator` or `SeriesFieldHelper`?

hugohaggmark commented on public/app/core/logs_model.ts:
```diff
@@ -326,3 +324,174 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+  const logSeries: SeriesData[] = [];
+
+  for (const series of seriesData) {
+    const logSeriesFieldProcessor = new LogsSeriesFieldProcessor(series);
```
Maybe we could new up this once and then process the series in the loop?

marefr replied:
You mean so that it can be used in line 382 as well, i.e. use an array of processors instead of looping seriesData?

hugohaggmark replied:
Well I was thinking of new:ing up objects in a loop, don't know how many objects there are in this case. As the rest of the discussion seemed to be leaned towards performance I added this. No biggie.

marefr replied:
Maximum 28 or something I suppose - the maximum of `refId`, a b c etc.

marefr replied:
This is now gone :boom:

hugohaggmark commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -85,29 +77,23 @@ export class LokiDatasource {
-      } else {
-        return { data: allStreams };
-      }
+      return { data: series };
```
Awesome!

hugohaggmark reviewed (approved):
Awesome work, found some very minor stuff.

marefr committed:
move IndexedField to processSeriesData

marefr commented:
> The name on `SeriesFieldsProcessor` feels a bit off, and potentially in conflict with some series processing/transforming work I hope to standardize in #16756
> 
> As implemented it is a way to find field indexes based on name/type -- maybe `FieldIndexFinder` or `IndexFieldFinder`?

@ryantxu I agree, I'm not super happy with that name. But it's not only a way for finding field indexes based on name/type - it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again. And for `LogsSeriesFieldProcessor` it adds some extra field type processing to find log level field => the log level finding seems kind of similar to your proposed matchers.

Would it be possible/a good thing to wrap the SeriesData in some sort of class/structure, similar to `SeriesFieldsProcessor`, mixed with your matchers, extensions and transforms and then use this class/structure internally in Grafana instead of SeriesData? 

I guess it's easier to start with renaming `SeriesFieldsProcessor` and get this merged and later if we merge your proposed PR and potential additional matchers/transforms we may be able to easier spot what is missing/how to properly refactor things.

marefr committed:
cleanup SeriesFieldProcessor

marefr commented:
@ryantxu I think changes in https://github.com/grafana/grafana/pull/16793/files#diff-317237f74791e054ee61a4e01ed3cb3bR58 and https://github.com/grafana/grafana/pull/16793/files#diff-c10a219f8d98d5afcc02054eb27f23a3R36 highlights the need of using some abstraction on top of `SeriesData`. What do you think?

ryantxu commented:
@marefr -- maybe, I don't think we need more now. regardless, I the details of this class should not hold up merging the PR.

ryantxu commented:
>  it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again.

FYI - this step already happens if you query with the new [PanelQueryRunner](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L347) right now that is only react panels, but soon will be everything.  Essentially it makes sure there is a type on everything before sending it to any panel.

marefr committed:
Merge remote-tracking branch 'origin/master' into 16287_loki_series_data

marefr commented:
> > it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again.
> 
> FYI - this step already happens if you query with the new [PanelQueryRunner](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L347) right now that is only react panels, but soon will be everything. Essentially it makes sure there is a type on everything before sending it to any panel.

I totally see the win and simplicity of using this. Next step for me in explore is to refractor how queries are executed and then I think it will make sense to use the same, `guessFieldTypes(toSeriesData(...))`, and later on add some additional field processing specific for logs. Then the `SeriesDataProcessor` may not be needed or it will be just a `FieldIndexFinder` or whatever we decide to call it. I'll give this a try tomorrow before merging.

ryantxu commented:
If you get ambitions and subscribe to the results, you may get streaming to work for "free" -- see:
https://github.com/grafana/grafana/pull/16815

marefr committed:
Merge remote-tracking branch 'origin/master' into 16287_loki_series_data

marefr committed:
refactor SeriesFieldsProcessor to FieldCache

Renames SeriesFieldsProcessor to FieldCache which makes it explicit to only be used for
lookup of fields by type or name.
Removes LogsSeriesFieldProcessor and postpone finding log level fields to later.

marefr committed:
explore: guess field types and convert to series data before processing logs

ryantxu commented:
@marefr -- looks great.  this is a huge improvement.

marefr commented:
Thanks for great feedback @ryantxu :+1: Really love how the new SeriesData format simplifies a lot of stuff! Next up is #16823 and next after that simplifying query execution in explore so that SeriesData is used for all both metrics and logs.

davkal commented:
@marefr using indexes as React keys is rarely advisable. The key tells react what to unmount, so now when you look at a loki stream (seeing all results) and then decide to add a search term that would match some of the later results (e.g., a term from row 10) and then run the query, the non-matched results still show. This happens because in the filtered set the first indexes are still present, so react thinks it won't have to unmount those rows.

davkal commented:
Opened issue: #16832
